Fail to decode string encoded as bin type

**Problem** 
msgp fails to decode string which were encoded as 'bin' type. It fails with the error: *msgp: too few bytes left to read object*

**Steps to reproduce:**

The following program generates the said error.

Note: The byte buffer is a valid msgpack buffer generated using embedded python program. Notice the use of "use_bin_type=True". When false, this error does not occur

```go
package main

import (
	"fmt"
	"os"
)


var buf = []byte{130, 196, 1, 97, 1, 196, 1, 98, 2}
// The above byte buffer was generated using the following python program
/*
import msgpack

m = {
		'a': 1,
		'b': 2,
}

buf = msgpack.dumps(m, use_bin_type=True)
print "[]byte{%s}" % ",".join([str(ord(n)) for n in buf])
*/

//go:generate msgp

type M struct {
	A int `msgp:"a"`
	B int `msgp:"b"`
}

func main() {
	m := &M{}
	_, err := m.UnmarshalMsg(buf[:])
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	fmt.Printf("%v\n", m)
}

```

I tried the `func TestReadMapKeyZC(t *testing.T)` and it does fail with the same reason:

```bash
--- FAIL: TestReadMapKeyZC (0.00s)
    a_test.go:13: msgp: too few bytes left to read object
```

--- a/msgp/read_bytes.go
+++ b/msgp/read_bytes.go
@@ -201,14 +201,14 @@ func ReadMapHeaderBytes(b []byte) (sz uint32, o []byte, err error) {
 // - ErrShortBytes (too few bytes)
 // - TypeError{} (not a str or bin)
 func ReadMapKeyZC(b []byte) ([]byte, []byte, error) {
-	o, b, err := ReadStringZC(b)
+	o, x, err := ReadStringZC(b)
 	if err != nil {
 		if tperr, ok := err.(TypeError); ok && tperr.Encoded == BinType {
 			return ReadBytesZC(b)
 		}
 		return nil, b, err
 	}
-	return o, b, nil
+	return o, x, nil
 }
 
 // ReadArrayHeaderBytes attempts to read

--- a/msgp/read_bytes_test.go
+++ b/msgp/read_bytes_test.go
@@ -552,6 +552,44 @@ func TestReadComplex128Bytes(t *testing.T) {
 	}
 }
 
+// both the 'str' and 'bin' types are acceptable map keys
+func TestReadMapKey(t *testing.T) {
+	args := []string{
+		"a",
+		"ab",
+		"qwertyuiop",
+	}
+	var buf bytes.Buffer
+	en := NewWriter(&buf)
+
+	for i := range args {
+		en.WriteString(args[i])
+		en.WriteBytes([]byte(args[i]))
+	}
+	en.Flush()
+	b := buf.Bytes()
+
+	for i := range args {
+		var s0, s1 []byte
+		var err error
+
+		s0, b, err = ReadMapKeyZC(b)
+		if err != nil {
+			t.Fatalf("couldn't read string as map key: %q", err)
+		}
+		s1, b, err = ReadMapKeyZC(b)
+		if err != nil {
+			t.Fatalf("couldn't read bytes as map key: %q", err)
+		}
+		if !bytes.Equal(s0, s1) {
+			t.Fatalf("%q != %q", s0, s1)
+		}
+		if string(s0) != args[i] {
+			t.Fatalf("%q != %q", s0, args[i])
+		}
+	}
+}
+
 func TestReadComplex64Bytes(t *testing.T) {
 	var buf bytes.Buffer
 	en := NewWriter(&buf)

